sw/device/tests/i2c_host_clock_stretching_test.c

```diff
@@ -0,0 +1,255 @@
+#include "sw/device/lib/base/macros.h"
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_i2c.h"
+#include "sw/device/lib/dif/dif_rv_plic.h"
+#include "sw/device/lib/runtime/hart.h"
+#include "sw/device/lib/runtime/irq.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/runtime/print.h"
+#include "sw/device/lib/testing/i2c_testutils.h"
+#include "sw/device/lib/testing/rv_core_ibex_testutils.h"
+#include "sw/device/lib/testing/rv_plic_testutils.h"
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+#include "i2c_regs.h"  // Generated.
+
+static_assert(__BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__,
+              "This test assumes the target platform is little endian.");
+
+OTTF_DEFINE_TEST_CONFIG();
+
+enum {
+
+  kHart = kTopEarlgreyPlicTargetIbex0,
+
+  kDeviceAddr = 0x22,
+  // Registers
+  kSwResetReg = 0xDF,
+  kManufacturerIdReg = 0xDE,
+  kProductIdReg = 0xDD,
+  kRxDelayReg = 0xDC,
+  kTxDelayReg = 0xDB,
+  kCache63BitsReg = 0xCF,
+  // Registers values
+  kManufacturerId = 0xA1,
+  kProductId = 0xA2,
+};
+
+/**
+ * Declared volatile because it is referenced in the main program flow as well
+ * as the ISR.
+ */
+// Hold the test result.
+static volatile status_t test_result;
+// Used to sync the irs and the main thread.
+static volatile dif_i2c_irq_t irq_fired;
+static dif_rv_plic_t plic;
+static dif_rv_core_ibex_t rv_core_ibex;
+static dif_pinmux_t pinmux;
+static dif_i2c_t i2c;
+
+/**
+ * Provides external IRQ handling for this test.
+ *
+ * This function overrides the default OTTF external ISR.
+ *
+ * For each IRQ, it performs the following:
+ * 1. Claims the IRQ fired (finds PLIC IRQ index).
+ * 2. Checks that the index belongs to the expected peripheral.
+ * 3. Checks that only the correct / expected IRQ is triggered.
+ * 4. Clears the IRQ at the peripheral.
+ * 5. Completes the IRQ service at PLIC.
+ */
+static status_t external_isr(void) {
+  dif_rv_plic_irq_id_t plic_irq_id;
+  TRY(dif_rv_plic_irq_claim(&plic, kHart, &plic_irq_id));
+
+  top_earlgrey_plic_peripheral_t peripheral = (top_earlgrey_plic_peripheral_t)
+      top_earlgrey_plic_interrupt_for_peripheral[plic_irq_id];
+  TRY_CHECK(peripheral == kTopEarlgreyPlicPeripheralI2c0,
+            "IRQ from incorrect peripheral: exp = %d(i2c0), found = %d",
+            kTopEarlgreyPlicPeripheralI2c0, peripheral);
+
+  irq_fired =
+      (dif_i2c_irq_t)(plic_irq_id - (dif_rv_plic_irq_id_t)
+                                        kTopEarlgreyPlicIrqIdI2c0FmtThreshold);
+
+  LOG_INFO("%s: plic:%d, i2c:%d", __func__, plic_irq_id, irq_fired);
+  TRY(dif_i2c_irq_acknowledge(&i2c, irq_fired));
+
+  // Complete the IRQ at PLIC.
+  TRY(dif_rv_plic_irq_complete(&plic, kHart, plic_irq_id));
+  return OK_STATUS();
+}
+
+void ottf_external_isr(void) { test_result = external_isr(); }
```

Comment:
Each IRQ will clobber the previous value of test result.  This should probably be something like:
```
tmp = external_isr();
if (status_ok(irq_result))
    irq_result = tmp;
```
This will record the first not-ok IRQ result and preserve it to the end of the test.